### ISSUE TITLE
`test/tests/weakref4.py` doesn't pass for me

### ISSUE THREAD

**Issue by tjhance** [dropbox/pyston#587]
```
$ ./pyston_dbg test/tests/weakref4.py
Traceback (most recent call last):
  File "test/tests/weakref4.py", line 31, in <module>:
    assert w() is None, "object was not collected"
AssertionError: object was not collected
```

**Comment by undingen** [comment]
Currently this tests passes for me but I run into the same issue after I tried to fix the "generator won't get freed when the don't stop" issue. While this won't help you :-( I'm quite happy to hear that this is may not be a bug I have introduced in my branch but instead the test is dependent on stack layout etc.

**Comment by toshok** [comment]
I ran into this as well, and duplicated the line that calls the function to
clear the stack. Problem went away.

Interestingly if you run it with -n the fix isn't required.

Sent from my iPhone

On Jun 5, 2015, at 12:03 AM, Travis Hance <notifications@github.com> wrote:

$ ./pyston_dbg test/tests/weakref4.py
Traceback (most recent call last):
  File "test/tests/weakref4.py", line 31, in <module>:
    assert w() is None, "object was not collected"
AssertionError: object was not collected

—
Reply to this email directly or view it on GitHub
<https://github.com/dropbox/pyston/issues/587>.

**Comment by kmod** [comment]
Hmm I feel like if the current hack in the test (the `recurse()` function) isn't enough to avoid these issues, there's probably something else going on that's worth looking into?

**Comment by toshok** [comment]
Agreed. It's also similar to an issue rudi's been investigating. Actually
it's probably the same issue, except instead of weakrefs its finalization.

On Jun 5, 2015, at 12:59 PM, Kevin Modzelewski <notifications@github.com>
wrote:

Hmm I feel like if the current hack in the test (the recurse() function)
isn't enough to avoid these issues, there's probably something else going
on that's worth looking into?

—
Reply to this email directly or view it on GitHub
<https://github.com/dropbox/pyston/issues/587#issuecomment-109421646>.

**Comment by kmod** [comment]
I'm going to tentatively say that this issue is solved by #590; if it happens again feel free to reopen.

That PR was just a workaround for the fact that we basically have no way of letting users guarantee the destruction of their objects.  Eventually we can move to precise stack scanning which will help with that a lot, though we will still have conservative scanning of extension objects and memory so I guess it will always remain a possibility.
